doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+---------------------------
+
+.. note::
+
+    See our AutoGraph guide for converting native Python control flow
+    to QJIT compatible control.
+
+There are various constraints and restrictions that should be kept in mind
+when working with classical control in Catalyst.
+
+- The return values of all branches of :func:`~.cond` must be the same type.
+  Returning different types, or ommitting a return value in one branch (e.g.,
+  returning ``None``) but not in others will result in an error.
+
+  >>> @qjit
+  ... def f(x: float):
+  ...     @cond(x > 1.5)
+  ...     def cond_fn():
+  ...         return x ** 2  # float
+  ...     @cond_fn.otherwise
+  ...     def else_branch():
+  ...         return 6  # int
+  ...     return cond_fn()
+  TypeError: Conditional requires consistent return types across all branches, got:
+  - Branch at index 0: [ShapedArray(float64[], weak_type=True)]
+  - Branch at index 1: [ShapedArray(int64[], weak_type=True)]
+  Please specify an else branch if none was specified.
+  >>> @qjit
+  ... def f(x: float):
+  ...     @cond(x > 1.5)
+  ...     def cond_fn():
+  ...         return x ** 2  # float
+  ...     @cond_fn.otherwise
+  ...     def else_branch():
+  ...         return 6.  # float
+  ...     return cond_fn()
+  >>> f(1.5)
+  array(6.)
+
+- Similarly, the else (``my_cond_fn.otherwise``) may be omitted **as long as
+  other branches do not return any values**. If other branches do return values,
+  the else branch must be specified.
+
+  >>> @qjit
+  ... def f(x: float):
+  ...     @cond(x > 1.5)
+  ...     def cond_fn():
+  ...         return x ** 2
+  ...     return cond_fn()
+  TypeError: Conditional requires consistent return types across all branches, got:
+  - Branch at index 0: [ShapedArray(float64[], weak_type=True)]
+  - Branch at index 1: []
+  Please specify an else branch if none was specified.
+
+  >>> @qjit
+  ... def f(x: float):
+  ...     @cond(x > 1.5)
+  ...     def cond_fn():
+  ...         return x ** 2
+  ...     @cond_fn.otherwise
+  ...     def else_branch():
+  ...         return x
+  ...     return cond_fn()
+  >>> f(1.6)
+  array(2.56)
+
+- Finally, a reminder that conditional functions provided to :func:`~.cond` cannot
+  accept any arguments.
+
+
+Compatibility with PennyLane transforms
+---------------------------------------
+
+PennyLane provides a wide variety of
+:doc:`transforms <code/qml_transforms>` that
+convert a circuit to one or more circuits.
+
+As a general rule of thumb, transforms that result in a single circuit are
+generally applied before the QNode decorator, while transforms that result in
+multiple circuits to be executed (**batch transforms**, such as
+:func:`~pennylane.transforms.split_non_commuting`) are applied after the QNode decorator.
+
+Currently, batch transforms and transforms that apply *after* the QNode decorator will
+not work with Catalyst:
+
+.. code-block:: python
+
+    @qjit
+    @qml.transforms.split_non_commuting
+    @qml.qnode(dev)
```

Comment:
I found the terms *before* and *after* the QNode decorator a bit confusing, because in the code they occur in opposite order than during execution.

Maybe *inside*/*outside* the qnode would be clearer, or some other terms? 🤔